feat / Dolomite DEX Integration

zrubenst opened the pull request:
Integrates the Dolomite Decentralized Exchange (https://dolomite.io) as a HummingBot exchange connector.

*Testing*
```
export WEB3_PROVIDER=<URL>
export DOLOMITE_TEST_PK=<PRIVATE_KEY>
export DOLOMITE_TEST_ADDR=<ADDRESS>
nosetests test/integration/test_dolomite_order_book_tracker.py
nosetests test/integration/test_dolomite_market.py
```

carlolm commented:
test this please

martinkou commented:
Hmm... this is a pretty major patch. I'm gonna need to take some time to test this myself.

It's nice to see you guys have included unit test cases.

coreycaplan3 commented:
:)

zrubenst commented:
@martinkou Updated the PR, removed the changes you requested be removed, and added support for interacting with our testnet (via setting a flag on the constructor for our MarketBase implementation). Our testnet is live at `testnet.dolomite.io`, you can connect with MetaMask by setting your custom RPC to `https://eth-testnet.dolomite.io` with a networkId of `1001`. At the moment, our connector is initialized by connecting to our testnet from the hummingbot_application.py. I figure once this PR is looked through and tested on your end we can change it to point to production.

martinkou commented on hummingbot/core/data_type/order_book_row.py:
```diff
@@ -1,21 +1,10 @@
     price: float
     amount: float
     update_id: int
     order_count: int
```
Please remove the `order_count` type annotation. It's no longer needed.

martinkou commented on hummingbot/market/dolomite/dolomite_market.pyx:
```diff
@@ -130,6 +158,14 @@ cdef class DolomiteMarket(MarketBase):
         self._in_flight_orders = {}
 
 
+    # @classmethod
```
Why is it commented out? It looks correct to me.

martinkou commented:
I'm not able to connect with the `https://eth-testnet.dolomite.io/` RPC endpoint to carry out tests. I cannot resolve an IP for the hostname `eth-testnet.dolomite.io`. Can you take a look why it's not working?

martinkou commented:
Oh, I just learned the correct node URL is `http://eth.dolomite.io/testnet`. That one works.

I'm gonna need some test ETH to perform my own tests though. How do I get test ETH on the testnet?

zrubenst commented:
@martinkou Sorry about that, we changed the RPC URL very recently, looks like you figured it out. If you login to the testnet exchange (testnet.dolomite.io) and get your account to tier 2 with a phone number (you can enter a fake home address on test net)we airdrop you 0.1 testnet ETH, 1250 testnet WETH and 250,000 testnet DAI for testing purposes.

zrubenst commented:
Also, this PR is incorrectly setting approvals at the moment, so in the meantime you can set the approval by using metamask in the testnet exchange to submit a buy and sell order (approvals are requested through metamask when you go to submit these at testnet.dolomite.io)

martinkou commented:
ok, I'll make some time tomorrow to test it again.

martinkou commented:
I tried both the sign up button and the "refill balances" button, none of them worked. So I'm still waiting for test ETH to perform testing.

Here's the error prompt related to sign up flow
<img width="2038" alt="Screen Shot 2019-09-27 at 1 39 44 PM" src="https://user-images.githubusercontent.com/304402/65800910-94a28400-e12c-11e9-8007-cf32dc3c20e8.png">

Here's a 404 error related to the Refill Balance button
<img width="1751" alt="Screen Shot 2019-09-27 at 1 40 52 PM" src="https://user-images.githubusercontent.com/304402/65800917-966c4780-e12c-11e9-8c1b-e60b5bd3c8d1.png">

martinkou commented:
Could you guys just send me some test ETH to 0x0717823d4a4aF7B5dF6855aa468712f5c15C9f75 so I can start testing?

mifeng commented:
We've merged the float ==> Decimal conversion to all exchanges that I mentioned to Cory last Thursday. Can you please make the corresponding changes to this PR?

coreycaplan3 commented:
Sorry it looks like we missed these comments. We restart the chain regularly because it's pretty unstable (we use Ganache). Balances should automatically refill, so I'll check and see what happened. It's like ely that you visited the testnet at a time in which we were between deployments or working through chain-related problems.

Surprisingly, all of these problems are non-issues for the mainnet. I'll make sure you ETH wallet has a balance now

coreycaplan3 commented:
@martinkou I also checked. The reason why you were 404'ing on the refill is because that wallet wasn't registered in our system. To my knowledge, any issues you faced with sign up should be fixed, so I'd encourage you to give it a shot tomorrow. 

It should fill your balance with ETH. It's not that simple in that we can send ETH to the wallet you provided because we have to restart the chain every day (one of the issues with running it via Ganache). Thus, we built a system that automatically funds wallets that are created/registered.

yzhang1994 commented:
test this please

yzhang1994 commented on hummingbot/client/settings.py:
```diff
@@ -50,6 +52,7 @@
     "radar_relay": "ZRX-WETH",
     "bamboo_relay": "ZRX-WETH",
     "coinbase_pro": "ETH-USDC",
+    "dolomite": "WETH-DAI"
```
missing comma

martinkou commented:
ok, I've managed to register an account in the testnet site tonight. Will conduct some testing tomorrow.

Thanks!

zrubenst commented on hummingbot/client/hummingbot_application.py:
```diff
@@ -307,6 +308,15 @@ def _initialize_markets(self, market_names: List[Tuple[str, List[str]]]):
+                    ethereum_rpc_url=ethereum_rpc_url,
+                    order_book_tracker_data_source_type=OrderBookTrackerDataSourceType.EXCHANGE_API,
+                    symbols=symbols,
+                    isTestNet=True,  # TODO: remove this! This needs to be False for release
```
This needs to be set to False when it is time to merge this into production

martinkou commented:
I still need a bit more time testing this... apparently I also need to modify the rest of hummingbot a bit to accommodate the different WETH and DAI contract addresses on the Dolomite test network.

zrubenst commented:
> I still need a bit more time testing this... apparently I also need to modify the rest of hummingbot a bit to accommodate the different WETH and DAI contract addresses on the Dolomite test network.

Okay! Just an FYI: I ran into the same issue and what I did was leave the RPC url to the one I use for mainnet. Dolomite's connector only relies on the RPC connection for setting approvals, every other interaction is performed through hummingbot hitting our API, and our server using it's own RPC connection (if hitting the testate API, it'll be the testnet RPC). What you can do is have hummingbot approve the testnet spender address on the mainnet (real) WETH/DAI token contracts. 

You can do this safely as the spender address on the mainnet is 1) not a contract and 2) the address that creates the contract on our testnet is the Ganache default wallet, which has made many main-chain transactions (presumably due to people sending tokens on mainnet when they meant to send them on a testnet) and thus the nonce that created the spender contract on our testnet has already been used on mainnet -- the spender address you set the approval for on mainnet cannot be used to steal your funds as a contract can never be deployed to that address.

If you want to refactor some of the code on hummingbot to make it work for testnets in general than ignore this solution, otherwise this is what I've been doing and it is safe. Especially if you use a burner wallet for testing this.

coreycaplan3 commented:
Just want to bump this and see where we're at.

martinkou commented:
I did some experiment myself to try to run pure market making strategy w/ Dolomite testnet with this patch: https://github.com/CoinAlpha/hummingbot/commit/867510a0636cf4871e802a57e55206c51a213285

The patch allows me to override the Ethereum chain ID and token addresses for WETH and DAI via hidden options in `conf_global.yml`.

Unfortunately, I wasn't able to run the strategy successfully in this way. I wasn't able to run the token approval transactions with the WETH and DAI smart contracts on the test net for some reason.

Attached the error logs I got from today's test run.

[hummingbot_logs_2019-10-17-16-11-08.log](https://github.com/CoinAlpha/hummingbot/files/3741662/hummingbot_logs_2019-10-17-16-11-08.log)

martinkou commented:
The clock tick errors about empty order books or the error about `'NoneType' object has no attribute 'actions'` are bugs on our old code... some of which have been fixed. So those can be ignored.

coreycaplan3 commented:
@martinkou It looks like the account you're using on Hummingbot didn't have ETH. Was this wallet signed up on Dolomite with an account? Regardless, replace your wallet with this one: 
- PK: `1234567812345678123456781234567812345678123456781234567812345678`
- Corresponding Wallet: `0x8D7f03FdE1A626223364E592740a233b72395235`

coreycaplan3 commented:
Let me know when you run tests as well. I just restarted our testnet not too long ago. Unfortunately we have some severe infra issues on it (our node goes down every ~24 hours).

martinkou commented:
Ok, just tested pure market making with the `0x8D7f03FdE1A626223364E592740a233b72395235 ` wallet in test network. It works.

Could you guys merge with the current development branch and also add in https://github.com/CoinAlpha/hummingbot/commit/867510a0636cf4871e802a57e55206c51a213285? Once that's done I'll approve and merge.

coreycaplan3 commented:
Done! @martinkou

coreycaplan3 commented:
Make sure you set the `isTestnet` variable to false in the initializer before this gets merged (if it hasn't been done already; I haven't checked)!

@martinkou

martinkou commented:
ok to test

martinkou commented:
The testnet flag is still True at the moment. Maybe you could set it back to false in `hummingbot_application.py`?

It can work either way though. I can merge it tomorrow after your change, or I merge it first and then make another PR to set that to False.

carlolm commented:
test this please
